draft-duke-quic-load-balancers.md

```diff
@@ -505,6 +518,208 @@ reserved for config rotation {{config-rotation}}.
+conditions in which a no-shared-state retry service can safely operate.
+
+Retry tokens generated by the service MUST have the format below.
+
+~~~~~
+0                   1                   2                   3
+0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|0| ODCIL (7) |  Original Destination Connection ID (0..160)    |      
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|           Original Destination Connection ID (...)            |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                     Opaque Data (variable)                    |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+{: #nss-retry-service-token-format title="Format of non-shared-state retry service tokens"}
+
+The first bit of retry tokens generated by the service must be zero. The token
+has the following additional fields:
+
+ODCIL: The length of the original destination connection ID from the triggering
+Initial packet. This is in cleartext to be readable for the server, but
+authenticated later in the token.
+
+Original Destination Connection ID: This also in cleartext and authenticated
+later.
+
+Opaque Data: This data MUST contain encrypted information that allows the retry
+service to validate the client's IP address, in accordance with the QUIC
+specification. It MUST also encode a secure hash of the original destination
+connection ID field to verify that this field has not been edited.
+
+Upon receipt of an Initial packet with a token that begins with '0', the retry
+service MUST validate the token in accordance with the QUIC specification. It
+must also verify that the secure hash of the Connect ID is correct. If
+incorrect, the token is invalid.
+
+In active mode, the service MUST issue Retry packets for all Client initial
+packets that contain no token, or a token that has the first bit set to '1'. It
+MUST NOT forward the packet to the server. The service MUST validate all tokens
+with the first bit set to '0'. If successful, the service MUST forward the
+packet with the token intact. If unsuccessful, it MUST drop the packet.
+
+Note that this scheme has a performance drawback. When the retry service is in
+active mode, clients with a token from a NEW_TOKEN frame will suffer a 1-RTT
+penalty even though it has proof of address with its token.
+
+In inactive mode, the service MUST forward all packets that have no token or a
+token with the first bit set to '1'. It MUST validate all tokens with the first
+bit set to '0'. If successful, the service MUST forward the packet with the
+token intact. If unsuccessful, it MUST either drop the packet or forward it
+with the token removed. The latter requires decryption and re-encryption of the
+entire Initial packet to avoid authentication failure. Forwarding the packet
```

Comment:
I don't think you can remove the token. I believe, per spec, the OCID TP parameter is only included if there is a token in the Initial, and a client must validate that if it sent a packet with a token that the TP parameter is present and correct; and if it did not send one, that the TP is no present. This explicitly prevents ANY kind of tampering.

Reply:
The point isn't to allow the connection to succeed. In normal QUIC, the signal that the Retry is from a MITM is received at the client. By stripping out the token, we preserve that convention.

If we leave the token, the server will assume it's valid and proceed, which is wrong. If we drop the packet, that's not great, as "client hello was dropped" is already overloaded. If anything, stripping out the token could be  a MUST.

Reply:
Bottom line, per Transport spec, if you strip the token, the client MUST eventually fail the handshake because the correct TP won't end up being present.